lib/elixir/lib/keyword.ex

```diff
@@ -186,6 +186,193 @@ defmodule Keyword do
     :lists.foldl(fun, [], Enum.reverse(pairs))
   end
 
+  @doc """
+  Ensures the first argument is a `keyword` with the given
+  keys and default values.
+  The second argument must be a list of atoms, specifying
+  a given key, or tuples specifying a key and a default value.
+  If any of the keys in the `keyword` is not defined on
+  `values`, it raises an error.
+
+  See also: `validate/2`
+
+  ## Examples
+      iex> Keyword.validate!([], [one: 1, two: 2]) |> Enum.sort()
+      [one: 1, two: 2]
+      iex> Keyword.validate!([two: 3], [one: 1, two: 2]) |> Enum.sort()
+      [one: 1, two: 3]
+
+  If atoms are given, they are supported as keys but do not
+  provide a default value:
+
+      iex> Keyword.validate!([], [:one, two: 2]) |> Enum.sort()
+      [two: 2]
+      iex> Keyword.validate!([one: 1], [:one, two: 2]) |> Enum.sort()
+      [one: 1, two: 2]
+
+  Passing an unknown key raises:
+
+      iex> Keyword.validate!([three: 3], [one: 1, two: 2])
+      ** (ArgumentError) unknown key :three in [three: 3], expected one of [:one, :two]
+  """
+  def validate!(keyword, values) when is_list(keyword) and is_list(values) do
+    # We use two lists to avoid reversing/concatenating
+    # lists in the middle of traversals.
+    case validate!(keyword, values, [], []) do
+      {:ok, keyword} ->
+        keyword
+
+      error ->
+        keys =
+          for value <- values do
+            if is_atom(value) do
+              value
+            else
+              elem(value, 0)
+            end
+          end
+
+        case error do
+          {:badkey, key} ->
+            raise ArgumentError,
+                  "unknown key #{inspect(key)} in #{inspect(keyword)}, " <>
+                    "expected one of #{inspect(keys)}"
+
+          :badkey ->
+            raise ArgumentError,
+                  "expected a keyword list with keys #{inspect(keys)}, got: #{inspect(keyword)}"
+        end
+    end
+  end
+
+  @doc """
+  Ensures the first argument is a `keyword` with the given
+  keys and default values.
+  The second argument must be a list of atoms, specifying
+  a given key, or tuples specifying a key and a default value.
+
+  Possible return values:
+  * `{:ok, keyword}`: Success case. Returns the validated keyword with default values applied.
+  * `{:error, :not_a_keyword}`: Indicates that the validated value is not a keyword.
+  * `{:error, :invalid_expected_values}`: Indicates that the expected values list is invalid.
```

Comment:
I'm not happy with some of these error messages, especially "invalid_expected_values". Suggestions are welcome!

Reply:
perhaps something like this?

```
  @doc """
  Similar to `validate/2` but returns the keyword or raises an error.

  ## Examples

      iex> Keyword.validate!([two: 3], [one: 1, two: 2])
      [one: 1, two: 3]

      iex> Keyword.validate!([three: 3], [one: 1, two: 2])
      ** (ArgumentError) unknown key :three in [three: 3], expected one of [:one, :two]

  """
```

btw,  I'd define the `validate!/2` function _after_ `validate/2` given they'd show up in docs in that order

Reply:
for these particular cases, instead of returning an error tuple I'd raise. This would be consistent with e.g. Keyword.fetch:

```
iex(1)> Keyword.fetch!("not a keyword", :key)
** (FunctionClauseError) no function clause matching in Keyword.fetch!/2

    The following arguments were given to Keyword.fetch!/2:

        # 1
        "not a keyword"

        # 2
        :key

    Attempted function clauses (showing 1 out of 1):

        def fetch!(keywords, key) when is_list(keywords) and is_atom(key)

    (elixir 1.13.0-dev) lib/keyword.ex:417: Keyword.fetch!/2
```

```
iex(2)> Keyword.fetch!([], "bad key")
** (FunctionClauseError) no function clause matching in Keyword.fetch!/2

    The following arguments were given to Keyword.fetch!/2:

        # 1
        []

        # 2
        "bad key"

    Attempted function clauses (showing 1 out of 1):

        def fetch!(keywords, key) when is_list(keywords) and is_atom(key)

    (elixir 1.13.0-dev) lib/keyword.ex:417: Keyword.fetch!/2
```

wdyt?

Reply:
Makes sense. I thought that maybe accepting invalid inputs would be desirable as well, but this makes for a better interface